### ISSUE TITLE
Questions about the Attention Map in Figure 5 of the paper

### ISSUE THREAD

**Issue by Sarimitsu0401** [bakqui/ST-MEM#8]
Dear author,

I very much appreciate the publication of your model.

If possible, could you please publish the code you used to draw Figure 5 in your paper?
I need it for interpretation of the model after fine tuning and would appreciate it.

Thanks.

**Comment by YunwonTae** [comment]
You may look at this repository, which we also referenced for our experiments.
https://github.com/naver-ai/cl-vs-mim/tree/main

**Comment by Sarimitsu0401** [comment]
Thank you very much.

For the pre-training model, I have drawn the averaged attention of the layers and the head with the explanation of figure5, but the result is not the same, and the first of each lead is the one we are interested in.

Is figure5 drawing a diagram with layer=10 and head=0 as in the repository you sent us?

**Comment by bakqui** [comment]
The (min-max normalized) attention weights were averaged across the layers and the heads. Please note that for Figure 5, the ECGs of the pretraining dataset (Chapman dataset specifically) were used.

**Comment by Sarimitsu0401** [comment]
Thank you very much.
I am very sorry. Please allow me to continue.
I think the final attention will be a [batch,384,384] matrix.
Could you please tell me what row of this matrix you are using?
(I was using row 0).
I am also converting it to 384 ->[12,32] then omitting the first and last elements on the column side to [12,30], and then converting it to [12,2250].
Is this the same for the author?

**Comment by bakqui** [comment]
1. Does "the final attention" means the averaged (w.r.t. the layers and the heads) attention scores?
If the "row" means the index number of the batch, it can be any number. We verified in the paper (Figure 8 in Appendix B.8) that the the attention map visulizations were similar regardless of the rhythm type of ECGs.

2. Yes. We dropped the the attention weights of [SEP] embeddings and made `(384, )` to `(12, 30)`.

**Comment by Sarimitsu0401** [comment]
Sorry for my poor English.
1. the final atteniton is, as you say, the averaged atteniton.
row is not a batch, but a query patch.

Also, in your paper you use the lead3 query patch, why do you use lead3?

2. thank you very much.

**Comment by bakqui** [comment]
> 1. the final atteniton is, as you say, the averaged atteniton.
>    row is not a batch, but a query patch.

You need to be careful to choose the query row. The size of the final attention would be `(384, 384)`. If you select the row as `0`, then the query would be the [SEP] embedding, which is not appropriate. If you would like to choose the first patch of lead I as a query, the row number needs to be `1`. We chose the row number as `32*2 + 8 = 71` for the Figure 5.

>Also, in your paper you use the lead3 query patch, why do you use lead3?

It is just for the visual highlights. We simply wanted to highlight that the model attended more to the spatially neighboring leads (e.g., III <-> aVF) and the similar morphology of the query.

**Comment by dduddus** [comment]
You may also try other samples or leads which were also depicted meaningful visualizations!

**Comment by Sarimitsu0401** [comment]
Sorry for the delay in replying.
Thank you for everything.

**Comment by Sarimitsu0401** [comment]
I would appreciate your input on this.
I am interested in which LEAD positions the model focuses on throughout.
Do you think the author is right about taking the average in the query patch to find this out?

**Comment by bakqui** [comment]
>I am interested in which LEAD positions the model focuses on throughout.
>Do you think the author is right about taking the average in the query patch to find this out?

Yes, I think it is appropriate.

**Comment by Sarimitsu0401** [comment]
Thank you so much for your help and for answering my question.
